File: _engine/typescript/v0.2.0/artifact.ts
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

import type { AIFuncArtifact } from './types';

export function validateArtifact(artifact: AIFuncArtifact): void {
  const name = artifact.name ?? artifact.package?.name;
  if (!name) {
    throw new Error('Artifact missing required field: name or package.name');
  }

  const hasEngine = artifact.engineVersion ?? artifact.package?.engine;
  if (!hasEngine) {
    throw new Error('Artifact missing required field: engineVersion or package.engine');
  }

  if (!artifact.prompt && !artifact.prompts) {
    throw new Error('Artifact missing required field: prompt or prompts');
  }

  if (!artifact.api) {
    throw new Error('Artifact missing required field: api');
  }
  if (!artifact.api.input) {
    throw new Error('Artifact missing required field: api.input');
  }
  if (!artifact.api.output) {
    throw new Error('Artifact missing required field: api.output');
  }
}

File: _engine/go/v0.2.0/artifact.go
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

package engine

import "fmt"

func ValidateArtifact(a AIFuncArtifact) error {
	name := a.Name
	if name == "" && a.Package != nil {
		name = a.Package.Name
	}
	if name == "" {
		return fmt.Errorf("artifact missing required field: name or package.name")
	}

	hasEngine := a.EngineVersion != ""
	if !hasEngine && a.Package != nil {
		hasEngine = a.Package.Engine != ""
	}
	if !hasEngine {
		return fmt.Errorf("artifact missing required field: engineVersion or package.engine")
	}

	if a.Prompt == "" && len(a.Prompts) == 0 {
		return fmt.Errorf("artifact missing required field: prompt or prompts")
	}

	if a.API.Input == nil {
		return fmt.Errorf("artifact missing required field: api.input")
	}
	if a.API.Output == nil {
		return fmt.Errorf("artifact missing required field: api.output")
	}

	return nil
}
